Intercept exceptions from the ServerRequest factory

weierophinney opened the pull request:
As noted in zendframework/zend-diactoros#171, the `ServerRequestFactory::fromGlobals()` method raises exceptions on malformed information, such as:

- malformed file upload data
- invalid HTTP protocol versions

At the application level, we should catch these, and invoke the final handler with a 400 response when caught.

I've opened this against master, as I consider the current behavior a "bug"; however, in a way, it's technically a new feature. I'd appreciate comments as to the preferred release. :smile:

weierophinney commented:
Ping @michaelmoussa — thoughts?

michaelmoussa commented on src/Application.php:
```diff
@@ -100,6 +103,13 @@ class Application extends MiddlewarePipe implements Router\RouteResultSubjectInt
+     *
+     * @var string
+     */
+    private $serverRequestFactory = ServerRequestFactory::class;
```
@weierophinney you _can_ mock this using Mockery without this private property + reflection if desired, by the way.

```php
/**
 * @runInSeparateProcess
 * @preserveGlobalState disabled
 */
public function testSomething()
{
    $serverRequestFactory = m::mock('alias:' . ServerRequestFactory::class);
    $serverRequestFactory->shouldReceive('fromGlobals')
        ->once()
        ->andReturn(['foo' => 'bar']);

    $this->assertSame(['foo' => 'bar'], ServerRequestFactory::fromGlobals());
}
```

Both annotations are required for this to work, else you will hit:

```
Mockery\Exception\RuntimeException: Could not load mock Zend\Diactoros\ServerRequestFactory, class already exists
```

No real harm in leaving it how it is now, but IMO the less "this is only used for testing" stuff in real classes, the better.

michaelmoussa commented:
@weierophinney PR looks good overall. Only suggestion was using Mockery alias mocks, which would allow us to remove an otherwise unnecessary private property as well as the two test asset classes.

Regarding bug vs feature, this looks like more of a "doing it this way is better than we are currently doing it" than "it is currently wrong and we must fix". Probably makes more sense for a minor release than a patch release.

weierophinney commented:
> Only suggestion was using Mockery alias mocks, which would allow us to remove an otherwise unnecessary private property as well as the two test asset classes.

I actually tried those first, but could not get them to work; I suspect the fact that the `ServerRequestFactory` was already autoloaded was getting in the way. One thing I did not try was running the test in a separate process, however; I'll give that a try shortly.

Regarding minor release vs patch, I'll rebase this against develop, and then update the PR to base it against there.

Thanks for the feedback, @michaelmoussa !

weierophinney commented:
Initial run worked, but then when I started removing the extraneous code, I started getting "serialization of closure is not allowed' errors from PHPUnit. Trying to debug...

weierophinney commented:
So, the solution was to wrap the test case in a try/catch block, which helped me discover that the response passed to the emitter was not the expected one... which brings me back to the original issue I had with `alias:`. I'll keep trying, though. I'd rather not have workarounds just for testing.

michaelmoussa commented:
@weierophinney "Serialization of Closure" in that context is usually due to what's being output by the test run in isolation. IIRC the tests in isolation communicate back to the main PHPUnit process by spitting out a serialized response which the main process reads.

If you get annoyed enough, feel free to share what you've got and I'll see if anything screams at me.

weierophinney commented:
@michaelmoussa Got it sorted; wrapped the internals in a try/catch, and the problem was with assertions in the closure failing. By wrapping in a try/catch, I was able to raise a failure assertion from the catch blocks.

The result: I discovered I wasn't importing `ServerRequestFactory`, which meant my double wasn't. :smile:

All is working now.

I rebased against develop, and squashed my commits so that we don't have a history of those unnecessary test assets, and then re-targeted this PR against the develop branch. Ready for review again!

danizord commented on test/ApplicationTest.php:
```diff
@@ -622,4 +626,64 @@ public function testPipingNotInvokableErrorMiddlewareRisesException()
+        $message
+    ) {
+        // try/catch is necessary in the case that the test fails.
+        // Assertion exceptions raised inside prophecy expectations normally
```
s/prophecy/mockery :P

danizord replied:
haaa nvm :D

michaelmoussa commented:
LGTM 👍

weierophinney commented:
@michaelmoussa I'll have you merge; I've got a few more proposals for 1.1.0 to sort out before we do a release, though, so don't tag. :)

michaelmoussa commented:
@weierophinney done. I used `feature/373` as the local merge branch rather than `hotfix` since you changed the base branch to `develop` and are planning to bundle in some other items before release. Hope that was correct.

weierophinney commented:
Perfect!

On Sep 2, 2016 12:00 AM, "Michael Moussa" <notifications@github.com> wrote:

> @weierophinney <https://github.com/weierophinney> done. I used feature/373
> as the local merge branch rather than hotfix since you changed the base
> branch to develop and are planning to bundle in some other items before
> release. Hope that was correct.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendframework/zend-expressive/pull/373#issuecomment-244283672>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABlV2z674fBBw8SMmuH-AkovytMs3Olks5ql618gaJpZM4JyFMQ>
> .
>
